src/main/java/bio/terra/service/dataset/DatasetDao.java

```diff
@@ -47,29 +54,143 @@ public DatasetDao(DataRepoJdbcConfiguration jdbcConfiguration,
+    /**
+     * Lock the dataset object before doing something with it (e.g. create, delete, bulk file load).
+     * This method returns successfully when there is a dataset object locked by this flight, and throws an exception
+     * in all other cases. Below is an outline of the logic flow of this method.
+     *     1. Check if the dataset is already locked.
+     *         a. If it's locked by THIS flight, then this must be a flight recovery, so return without error.
+     *         b. If it's locked by ANOTHER flight, then throw an exception.
+     *     2. Depending on the LockBehaviorFlag specified, throw an exception if the dataset does/not already exist.
+     *     3. Either create a new dataset or update the existing one, and give this flight the lock.
+     * @param datasetName name of the dataset to lock, this is a unique column
+     * @param defaultProfileId default profile id of the dataset, this is a required column for creating a new dataset
+     * @param flightId flight id that wants to lock the dataset
+     * @param lockFlag flag specifying the lock behavior (i.e. whether to throw an exception if dataset does/not exist)
+     * @throws DatasetLockException if the dataset is locked by another flight,
+     *                              or if it does/not exist according to the LockBehaviorFlag specified
+     */
+    @Transactional(propagation =  Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+    public void lock(String datasetName, UUID defaultProfileId, String flightId, LockBehaviorFlags lockFlag) {
+        if (flightId == null) {
+            throw new DatasetLockException("Locking flight id cannot be null");
+        }
+
+        // lookup the dataset
+        String sql = "SELECT flightid FROM dataset WHERE name = :name";
+        MapSqlParameterSource params = new MapSqlParameterSource().addValue("name", datasetName);
+        List<String> selectResult = jdbcTemplate.query(sql, params,
+            (rs, rowNum) -> rs.getString(1));
+
+        boolean datasetExists = (selectResult.size() == 1);
+        String foundFlightId = datasetExists ? selectResult.get(0) : null;
+        boolean datasetLocked = foundFlightId != null;
+
+        // if the dataset is already locked...
+        if (datasetLocked && flightId.equals(foundFlightId)) {
+            // ...by THIS flight, then this must be a flight recovery
+            logger.debug("dataset already locked by THIS flight, must be a flight recovery");
+            return; // return without error
+        } else if (datasetLocked && !flightId.equals(foundFlightId)) {
+            // ...by ANOTHER flight, then throw an exception
+            logger.debug("dataset already locked by ANOTHER flight, error");
+            throw new DatasetLockException("Dataset is locked by another flight: " + foundFlightId);
+        }
+
+        // at this point, we know that the dataset is not locked by anyone
+        logger.debug("dataset not locked by anyone");
```

Comment:
Is it possible that another flight locks the dataset the instant after the `datasetLocked` check above but before this line is hit?

Reply:
I believe the purpose of the @Transactional method annotation is to start/end a JDBC transaction before/after the method. The propagation and isolation specified in the method signature correspond to the concurrency and isolation properties of a JDBC transaction. So to avoid dirty reads, I set the isolation to serializable.
https://docs.spring.io/spring/docs/4.2.x/spring-framework-reference/html/transaction.html
https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/annotation/Isolation.html

Reply:
To rephrase, does this effectively ensure that the entire method runs transactionally?

Reply:
It ensures that the method is wrapped in a JDBC transaction. So it does not apply to anything other than JDBC statements. It has no effect on accessing Java structures in a thread-safe way. I don't follow why this is relevant for the locking though. The lock is a populated field in a JDBC row. So accessing the lock is done within a JDBC transaction.

Reply:
I think I understand how this works now. dd also helped me understand this better. Thanks!